### PR TITLE
replace checked_fptosi intrinsics with Julia implementation

### LINKED ISSUE #14549: convert(Int16, ::Float64) does not throw inexact exception when it should
```jl
julia> convert(Int16, 88776.0)
23240
```

LLVM IR (wrapped in another function)
```llvm
define i16 @julia_i16_convert_22530(double) #0 {
top:
  %1 = fptosi double %0 to i16
  %2 = sitofp i16 %1 to double
  %3 = fcmp oeq double %2, %0
  br i1 %3, label %pass, label %fail

fail:                                             ; preds = %top
  %4 = load %jl_value_t*, %jl_value_t** @jl_inexact_exception, align 4
  call void @jl_throw(%jl_value_t* %4)
  unreachable

pass:                                             ; preds = %top
  ret i16 %1
}
```

ASM
```asm
Dump of assembler code for function julia_i16_convert_22473:
   0xcce3b000 <+0>:     push    {r11, lr}
   0xcce3b004 <+4>:     mov     r11, sp
   0xcce3b008 <+8>:     vcvt.s32.f64    s2, d0
   0xcce3b00c <+12>:    vmov    r0, s2
   0xcce3b010 <+16>:    vcvt.f64.s32    d16, s2
   0xcce3b014 <+20>:    vcmpe.f64       d16, d0
   0xcce3b018 <+24>:    vmrs    APSR_nzcv, fpscr
   0xcce3b01c <+28>:    popeq   {r11, pc}
   0xcce3b020 <+32>:    movw    r0, #2920       ; 0xb68
   0xcce3b024 <+36>:    movt    r0, #63345      ; 0xf771
   0xcce3b028 <+40>:    ldr     r0, [r0]
   0xcce3b02c <+44>:    bl      0xcce3b050
End of assembler dump.
```

(interestingly this doesn't happen on AArch64)

### PR DESCRIPTION
**Pull request by simonbyrne**
Fixes #14549.

As @yuyichao points out, this will effect inlining rules.

### CODE REVIEW

**Comment by yuyichao** [comment]
Let's check =) `runbenchmarks(ALL, vs = "JuliaLang/julia:master")`

**Comment by simonbyrne** [comment]
@nanosoldier `runbenchmarks(ALL, vs=":master")`

**Comment by simonbyrne** [comment]
This has been updated to do an explicit range check, so should no longer utilise any undefined behaviour

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/14763#issuecomment-229201341) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/ef0fa232d9b50b6eba3dd46e84bdc70ee8ab0941/f7ab200_vs_0dcb477/report.md). cc @jrevels

**Comment by simonbyrne** [comment]
Hmm, the slowdown seems to be due to the fact that the length of `FloatRange` and `LinSpace` is a floating point number, which needs to be converted a lot.

File: base/float.jl [hunk 1]
```diff
@@ -463,16 +463,29 @@ for Ti in (Int8, Int16, Int32, Int64, Int128, UInt8, UInt16, UInt32, UInt64, UIn
...
             end
-        else
+            @eval function convert(::Type{$Ti}, x::$Tf)
+                if ($(Tf(typemin(Ti))) <= x <= $(Tf(typemax(Ti)))) && (trunc(x) == x)
+                    return unsafe_trunc($Ti,x)
+                else
+                    throw(InexactError())
+                end
+            end
+        else #
```
Comment (tkelman) [inline base/float.jl]: was there going to be a comment here then changed your mind?

Comment (simonbyrne) [inline base/float.jl]: I can't remember, this was quite a long time ago (if only I had left a comment on what that comment was going to be...)

**Comment by simonbyrne** [comment]
@nanosoldier runbenchmarks(ALL, vs=":master")

**Comment by jrevels** [comment]
@simonbyrne looks like you may have submitted the job with the wrong syntax, and then edited to the correct syntax? Nanosoldier ignores comment edits to prevent accidental job resubmission. Triggering it with a new comment should work:

@nanosoldier `runbenchmarks(ALL, vs=":master")`

**Comment by simonbyrne** [comment]
thanks!

**Comment by vchuravy** [comment]
Would this also work for Float16?

**Comment by simonbyrne** [comment]
It should do, the logic is pretty much the same.

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/14763#issuecomment-249230434) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/455c148664ba4c5c8eb8caab34aa4d89345032cb/9874dd7_vs_792ded8/report.md). cc @jrevels

**Comment by simonbyrne** [comment]
Actually, it won't work for `Float16`, but I've added a note as to why. I didn't include the Float16 logic here, as the current method (converting to Float32 first) is probably more efficient.

**Comment by simonbyrne** [comment]
Updated. It seems like the change in performance is a bit of a wash. We may want to look at speeding up `FloatRange` and `LinSpace` either by inlining or perhaps even redesigning so the length is stored as an integer, but this is probably better in a different PR.

Any objections to merging?

**Comment by vchuravy** [comment]
No objections to merging.

I would still be interested in a implementation for Float16. Mostly because I am looking at improving our support for Float16 on platforms that support them natively.

**Comment by simonbyrne** [comment]
It should be possible via an extra branch: just check if `Tf(typemin(Ti)) == Inf`, in which case use an inequality rather than equality for the lower bound.

**Comment by tkelman** [comment]
that's a pretty bad slowdown in sparse indexing, worth looking into and profiling the difference

**Comment by simonbyrne** [comment]
I couldn't figure it out: there don't appear to be any conversion calls, and I can't recreate it locally.

One more time:

@nanosoldier `runbenchmarks(ALL, vs=":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/14763#issuecomment-249389007) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/eb6a960b137d2fcc793aa3d74d8b5ac2294ffce1/415de41_vs_11b06ed/report.md). cc @jrevels

**Comment by simonbyrne** [comment]
The sparse issue seems to have mostly gone.

**Comment by simonbyrne** [comment]
Given the inconsistency, it seems that the regressions are mostly noise. We do get a nice speed boost on some though. Shall we merge this?

**Comment by musm** [comment]
Out of curiosity, how reliable is nanosoldier if +-15% can be mainly attributed to noise?

**Comment by simonbyrne** [comment]
@nanosoldier `runbenchmarks(ALL, vs=":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/14763#issuecomment-250694193) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/e456ed085730c22eedb80fd714f62d7e97f374b1/e3988a4_vs_c40090f/report.md). cc @jrevels

**Comment by simonbyrne** [comment]
Okay, we'll have to figure out how to improve `FloatRange` and `LinSpace` indexing, but I think this is good to go.

**Comment by simonbyrne** [comment]
Will probably want to backport this to get tests to pass on ARM/Power in next release.

**Comment by tkelman** [comment]
`jl_checked_fptoui` and `jl_checked_fptosi` were exported from libjulia, but doesn't look like any packages call into them

**Pull request merged**